Add spec changes for image extensions (#298)
buildpacks/spec: platform.md

The platform SHOULD ensure that:

- The image config's `User` field is set to a user with a **DIFFERENT** user [†](README.md#operating-system-conventions)UID/[‡](README.md#operating-system-conventions)SID as the build image.
- The image config's `Env` field has the environment variable `CNB_USER_ID` set to the user [†](README.md#operating-system-conventions)UID/[‡](README.md#operating-system-conventions)SID of the user specified in the `User` field.

> - The image config's `Env` field has the environment variable `CNB_GROUP_ID` set to the primary group [†](README.md#operating-system-conventions)GID/[‡](README.md#operating-system-conventions)SID of the user specified in the `User` field.

Author: For the lifecycle to know what value to provide as `user_id` and `group_id` as build args to the first Dockerfile

Author: Actually, this is not necessary since the lifecycle can read this information off of the image config in order to provide the build args. I'll remove it.

| `/dev/stderr`                                | Logs (warnings, errors)                                                       |
| `<output>/<image extension ID>/<Dockerfile>` | Generated Dockerfiles (see [Image Extension Specfication](image-extension.md) |
| `<layers>/config/metadata.toml`              | Build metadata (see [`metadata.toml`](#metadatatoml-toml))                    |

| Exit Code       | Result                              |
|-----------------|-------------------------------------|
| `0`             | Success                             |
| `11`            | Platform API incompatibility error  |
| `12`            | Buildpack API incompatibility error |
| `1-10`, `13-19` | Generic lifecycle errors            |
| `91`            | Image extension build error         |
| `90`, `92-99`   | Build-Ext-specific lifecycle errors |

- The lifecycle SHALL execute all image extensions in the order defined in `<group>` according to the process outlined in the [Buildpack Interface Specification](buildpack.md).
- The lifecycle SHALL add all invoked image extensions to`<layers>/config/metadata.toml`.
- The lifecycle SHALL copy all output Dockerfiles to `<output>`.

> - The lifecycle SHALL record all Dockerfiles and build args in `<layers>/extender.toml`.

Author: ```suggestion
- The lifecycle SHALL record all Dockerfiles and build args in `<extender>`.
```

> - The lifecycle SHALL filter the build plan for buildpacks with dependencies provided by image extensions.

Author: ```suggestion
- The lifecycle SHALL filter the build plan with dependencies provided by image extensions.
```

#### `extender`

The platform MAY execute `extender` for Linux builds

Usage:

```
/cnb/lifecycle/extender \
  [-analyzed <analyzed>] \
  [-app <app>] \
  [-cache-image <cache-image>] \
  [-extender <extender> ] \
  [-genpkgs <genpkgs> ] \
  [-ignore-paths <ignore-paths>] \
  [-kind <kind>] \
  [-log-level <log-level>] \
  [-work-dir <work-dir>] \
  <base-image> \

>   <target-image>

Author: This interface is a bit weird... I can't remember why it ended up this way. Perhaps better would be to accept either a `-build-image` or `-run-image` flag (from which the `kind` can be inferred) and default to `kind=run` when no flag is provided, reading the run image from analyzed.toml.

| `<genpkgs>`      | `CNB_GENPKS_PATH`         | `<work-dir>/genpkgs`                              | Path to a `genkpgs` binary                                           |
| `<ignore-paths>` | `CNB_EXTEND_IGNORE_PATHS` |                                                   | Comma separated list of paths to ignore                              |
| `<kind>`         |                           | `build`                                           | Type of base image to extend (valid values: `build`, `run`)          |
| `<log-level>`    | `CNB_LOG_LEVEL`           | `info`                                            | Log Level                                                            |
| `<target-image>` |                           |                                                   | Target image reference                                               |
| `<work-dir>`     | `CNB_EXTEND_WORK_DIR`     | `<layers>`                                        | Path to a working directory                                          |

- `<base-image>`, `<cache-image>`, and `<target-image>` MUST be valid image references

> - If after extending the image, the `User` defined on the image is `root`, the lifecycle MUST fail

Author: ```suggestion
- If after extending a base image, the `User` defined on the image is `root`, the lifecycle MUST fail
```

> - If after extending the run image, the underlying base image has not changed but the `User` defined on the image has changed from its original value, the lifecycle SHOULD warn

Author: ```suggestion
- If after extending the run image, the underlying base image has not changed but image config's `User` field has changed from its original value, the lifecycle SHOULD warn
```

When extending the run image:

> - After all image extensions have been applied, the lifecycle MUST set the label `io.buildpacks.rebasable` to `true` if and only if the original base image had `io.buildpacks.rebasable=true` and all image extensions set the label `io.buildpacks.rebasable` to `true`.

Author: ```suggestion
- After all image extensions have been applied, the lifecycle MUST set the label `io.buildpacks.rebasable` to `true` if and only if the original base image had `io.buildpacks.rebasable=true` and all Dockerfiles set the label `io.buildpacks.rebasable` to `true`.
```

> - After all image extensions have been applied, the lifecycle MUST run `<genpkgs>` (if provided) in the context of the extended run image to generate an SBOM for the extended run image. The lifecycle MUST include the SBOM as a layer in the extended run image. The diffID of the layer containing the run image SBOM SHOULD replace the `io.buildpacks.base.sbom` label on the extended run image.

Author: Basing this off of https://github.com/buildpacks/rfcs/pull/186 - but there is also https://github.com/buildpacks/rfcs/pull/211, which would also work. The platform would need to specify a location to write the generated SBOM.

      - All run-image layers SHALL be preserved
      - All run-image config values SHALL be preserved unless this conflicts with another requirement
    - MUST contain all buildpack-provided launch layers as determined by the [Buildpack Interface Specfication](buildpack.md)

>     - MUST contain a layer containing all image extension-provided and buildpack-provided Software Bill of Materials (SBOM) files for `launch` as determined by the [Buildpack Interface Specfication](buildpack.md) if they are present

Author: This is a bit weird. If image extension SBOM files are included here, there could be collisions between image extension ids and buildpack ids... I am still not sure when image extensions would need to output SBOM files, because we have `genpkgs`. It would be nice to remove this if we can.

Author: Per latest discussion, we decided to take SBOM files out of the extension api for now. An extension could use a Dockerfile to write an SBOM file for a downloaded artifact if necessary.

      - `<layers>/sbom/<buildpack-id>/launch.sbom.<ext>` MUST contain the buildpack-provided `launch` SBOM
      - `<layers>/sbom/<buildpack-id>/<layer-id>/launch.sbom.<ext>` MUST contain the buildpack-provided layer SBOM if `<layer-id>` is a `launch` layer

>       - TODO: if image extension SBOM files are included here, there could be collisions between image extension ids and buildpack ids...

Author: ```suggestion
```
